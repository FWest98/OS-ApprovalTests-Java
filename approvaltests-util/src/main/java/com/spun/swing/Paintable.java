package com.spun.swing;

import java.awt.Dimension;
import java.awt.Graphics;

public interface Paintable
{
  Dimension getSize();
  void paint(Graphics g);
}
