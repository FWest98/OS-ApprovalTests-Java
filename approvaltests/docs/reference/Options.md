<a id="top"></a>

# Options

<!-- toc -->
## Contents

* [Introduction](#introduction)
* [Fluent Interface](#fluent-interface)
* [Reporters](#reporters)
* [Scrubbers](#scrubbers)
* [File Options](#file-options)
    * [File Extensions](#file-extensions)
* [Defaults](#defaults)
* [Adding to Existing Options object](#adding-to-existing-options-object)<!-- endToc -->

## Introduction

There are many things you might want to tweak with Approval Tests. `Options` is the entry-point for many of the changes.
It is on all `verify()` methods, as an optional parameter.

## Fluent Interface

`Options` utilizes a fluent interface, allowing you to chain together commands. Each returned object is a new copy.

<!-- snippet: specify_all_the_options -->
<a id='snippet-specify_all_the_options'></a>
```cpp
ApprovalTests::Options()
    .withReporter(ApprovalTests::QuietReporter())
    .withScrubber(ApprovalTests::Scrubbers::scrubGuid)
    .fileOptions().withFileExtension(".json")
```
<sup><a href='/tests/DocTest_Tests/DocTestApprovalTestTests.cpp#L50-L55' title='Snippet source file'>snippet source</a> | <a href='#snippet-specify_all_the_options' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

## Reporters

[Reporters](/doc/Reporters.md#top) launch diff tools upon failure.  Read how to configure them with `Options` [here](../Reporters.md#via-options):

## Scrubbers

[Scrubbers](/doc/explanations/Scrubbers.md#top) clean output to help remove inconsistent pieces of text, such as dates.  Read how to set up Scrubbers [here](../Scrubbers.md#configuring-scrubbers):
There are two ways to set a Scrubber.

## File Options

The `Options::FileOptions` class exists to customise the `.approved` and `.received` files in various ways.

For now, it just controls the file extension.

### File Extensions

If you want to change the file extension of both the approved and received files, use `withFileExtension()`.

<!-- snippet: basic_approval_with_file_extension -->
<a id='snippet-basic_approval_with_file_extension'></a>
```cpp
using namespace ApprovalTests;

Approvals::verify("text to be verified",
                  Options().fileOptions().withFileExtension(".xyz"));
```
<sup><a href='/tests/DocTest_Tests/core/OptionsExamples.cpp#L7-L12' title='Snippet source file'>snippet source</a> | <a href='#snippet-basic_approval_with_file_extension' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

**Note:** `withFileExtension()` returns an `Options` object, so it's possible to keep appending more `with...()` calls.

## Defaults

The default constructor for Options does:

* no scrubbing
* uses file extension `.txt`
* uses whatever is currently set as [the default reporter](#registering-a-default-reporter).

## Adding to Existing Options object

Because of the fluent options, you can modify a specific option from an existing Options object,
while retaining all other settings, and not changing the original object.

```cpp
verifyWithQuietReporter(std::string text, const Options& o)
{
    Approvals::verify(text, o.withReporter(QuietReporter());
}
```

---

[Back to User Guide](/doc/README.md#top)