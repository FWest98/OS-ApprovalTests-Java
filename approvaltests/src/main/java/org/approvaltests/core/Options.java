package org.approvaltests.core;

import com.spun.util.ArrayUtils;
import org.approvaltests.Approvals;
import org.approvaltests.ReporterFactory;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.NamerWrapper;
import org.approvaltests.scrubbers.NoOpScrubber;
import org.approvaltests.writers.ApprovalWriterFactory;
import org.approvaltests.writers.DefaultApprovalWriterFactory;

import java.util.HashMap;
import java.util.Map;

public class Options
{
  private enum Fields {
                       SCRUBBER, REPORTER, FILE_OPTIONS_FILE_EXTENSION, FILE_OPTIONS_NAMER, WRITER;
  }
  private final Map<Fields, Object> fields = new HashMap<>();
  public Options()
  {
  }
  public Options(Scrubber scrubber)
  {
    fields.put(Fields.SCRUBBER, scrubber);
  }
  public Options(ApprovalFailureReporter reporter)
  {
    fields.put(Fields.REPORTER, reporter);
  }
  private Options(Map<Fields, Object> fields, Fields key, Object value)
  {
    this.fields.putAll(fields);
    this.fields.put(key, value);
  }
  public ApprovalFailureReporter getReporter()
  {
    return ArrayUtils.getOrElse(fields, Fields.REPORTER, ReporterFactory::get);
  }
  public Options withReporter(ApprovalFailureReporter reporter)
  {
    return new Options(fields, Fields.REPORTER, reporter);
  }
  public Options withScrubber(Scrubber scrubber)
  {
    return new Options(fields, Fields.SCRUBBER, scrubber);
  }
  public String scrub(String input)
  {
    return getScrubber().scrub(input);
  }
  private Scrubber getScrubber()
  {
    return ArrayUtils.getOrElse(fields, Fields.SCRUBBER, () -> NoOpScrubber.INSTANCE);
  }
  public ApprovalWriter createWriter(Object o)
  {
    ApprovalWriterFactory factory = ArrayUtils.getOrElse(fields, Fields.WRITER,
        DefaultApprovalWriterFactory::getDefaultFactory);
    return factory.create(o, this);
  }
  public Options withWriter(ApprovalWriterFactory approvalWriterFactory)
  {
    return new Options(fields, Fields.WRITER, approvalWriterFactory);
  }
  public FileOptions forFile()
  {
    return new FileOptions(this.fields);
  }
  public static class FileOptions
  {
    private final Map<Fields, Object> fields;
    public FileOptions(Map<Fields, Object> fields)
    {
      this.fields = fields;
    }
    public Options withExtension(String fileExtensionWithDot)
    {
      if (!fileExtensionWithDot.startsWith("."))
      {
        fileExtensionWithDot = "." + fileExtensionWithDot;
      }
      return new Options(fields, Fields.FILE_OPTIONS_FILE_EXTENSION, fileExtensionWithDot);
    }
    public ApprovalNamer getNamer()
    {
      return ArrayUtils.getOrElse(fields, Fields.FILE_OPTIONS_NAMER, Approvals::createApprovalNamer);
    }
    public String getFileExtension()
    {
      return ArrayUtils.getOrElse(fields, Fields.FILE_OPTIONS_FILE_EXTENSION, () -> ".txt");
    }
    public Options withBaseName(String fileBaseName)
    {
      NamerWrapper approvalNamer = new NamerWrapper(() -> fileBaseName, getNamer());
      return new Options(fields, Fields.FILE_OPTIONS_NAMER, approvalNamer);
    }
    public Options withName(String fileBaseName, String extension)
    {
      NamerWrapper approvalNamer = new NamerWrapper(() -> fileBaseName, getNamer());
      HashMap<Fields, Object> newFields = new HashMap<>(fields);
      newFields.put(Fields.FILE_OPTIONS_FILE_EXTENSION, extension);
      return new Options(newFields, Fields.FILE_OPTIONS_NAMER, approvalNamer);
    }
  }
}
