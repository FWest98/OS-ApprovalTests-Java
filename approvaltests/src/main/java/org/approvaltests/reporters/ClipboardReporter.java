package org.approvaltests.reporters;

import com.spun.util.SystemUtils;
import com.spun.util.WindowUtils;

import java.awt.GraphicsEnvironment;
import java.io.File;

/**
 * A getReporter which creates the command to accept the received
 * file as the approve file and copies it to the clipboard:<br>
 * <code>move received.txt approved.txt</code>
 */
public class ClipboardReporter implements EnvironmentAwareReporter
{
  @Override
  public void report(String received, String approved)
  {
    WindowUtils.copyToClipBoard(getCommandLine(received, approved), false);
  }
  public static String getCommandLine(String received, String approved)
  {
    File r = new File(received);
    File a = new File(approved);
    String commandLine = ClipboardReporter.getAcceptApprovalText(r.getAbsolutePath(), a.getAbsolutePath());
    return commandLine;
  }
  /**
   * The clipboard will not be available in a headless environment.
   */
  @Override
  public boolean isWorkingInThisEnvironment(String forFile)
  {
    return !GraphicsEnvironment.isHeadless();
  }
  public static String getAcceptApprovalText(String received, String approved)
  {
    return getAcceptApprovalText(received, approved, SystemUtils.isWindowsEnviroment());
  }
  public static String getAcceptApprovalText(String received, String approved, boolean windowsEnviroment)
  {
    if (windowsEnviroment)
    {
      return String.format("move /Y \"%s\"  \"%s\"", received, approved);
    }
    else
    {
      return String.format("mv -f %s %s", formatFilePathForCommandLine(received),
          formatFilePathForCommandLine(approved));
    }
  }
  private static String formatFilePathForCommandLine(String filePath)
  {
    if (filePath.contains(" "))
    { return '"' + filePath + '"'; }
    return filePath;
  }
}
