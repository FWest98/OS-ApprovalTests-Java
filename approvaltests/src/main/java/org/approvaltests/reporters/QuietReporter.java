package org.approvaltests.reporters;

import com.spun.util.logger.SimpleLogger;

public class QuietReporter implements EnvironmentAwareReporter
{
  public static final QuietReporter INSTANCE = new QuietReporter();
  @Override
  public boolean report(String received, String approved)
  {
    SimpleLogger.message(ClipboardReporter.getAcceptApprovalText(received, approved));
    return true;
  }
  @Override
  public boolean isWorkingInThisEnvironment(String forFile)
  {
    return true;
  }
}
