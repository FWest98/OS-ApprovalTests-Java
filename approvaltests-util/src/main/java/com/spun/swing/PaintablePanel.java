package com.spun.swing;

import java.awt.Graphics;
import javax.swing.JPanel;

public class PaintablePanel<P extends Paintable> extends JPanel
{
  private final P paintable;
  public PaintablePanel(P paintable)
  {
    this.paintable = paintable;
    setPreferredSize(paintable.getSize());
    paintable.registerRepaint(() -> this.repaint());
  }
  @Override
  public void paint(Graphics g)
  {
    paintable.paint(g);
  }
  public P get()
  {
    return paintable;
  }
}
