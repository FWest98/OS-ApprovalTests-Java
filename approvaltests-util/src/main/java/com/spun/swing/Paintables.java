package com.spun.swing;

import java.awt.Graphics;
import javax.swing.JPanel;

public class Paintables
{
  public static JPanel asPanel(Paintable paintable)
  {
    return new JPanel()
    {
      {
        setPreferredSize(paintable.getSize());
      }
      @Override
      public void paint(Graphics g)
      {
        paintable.paint(g);
      }
    };
  }
}
