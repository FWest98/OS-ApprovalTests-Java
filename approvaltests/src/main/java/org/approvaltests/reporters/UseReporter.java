package org.approvaltests.reporters;

import org.approvaltests.core.ApprovalFailureReporter;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@Retention(RetentionPolicy.RUNTIME)
public @interface UseReporter {
  Class<? extends ApprovalFailureReporter>[] value();
}
