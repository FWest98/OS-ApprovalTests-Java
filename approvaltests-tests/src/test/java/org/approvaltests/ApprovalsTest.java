package org.approvaltests;

import java.awt.Dimension;
import java.awt.Rectangle;

import org.approvaltests.core.Options;
import org.approvaltests.reporters.ClipboardReporter;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.QuietReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.Test;

import javax.swing.JButton;

@UseReporter({DiffReporter.class, ClipboardReporter.class})
public class ApprovalsTest
{
  @Test
  public void testToString() throws Exception
  {
    Approvals.verify(new Rectangle(5, 10, 100, 200));
  }
  @Test
  public void testAsJson() throws Exception
  {
    Approvals.verifyAsJson(new Rectangle(5, 10, 100, 200));
  }
  @Test
  public void testWrongCall()
  {
    Approvals.verifyException(() -> {
      JButton b = new JButton("Approval Tests Rule");
      b.setPreferredSize(new Dimension(150, 20));
      Approvals.verify(b, new Options(new QuietReporter()));
    });
  }
}
