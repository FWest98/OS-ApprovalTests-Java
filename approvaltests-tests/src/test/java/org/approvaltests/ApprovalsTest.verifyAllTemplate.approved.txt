TITLE


placeholder input.value1
placeholder input.value2
